### ISSUE TITLE
Unwrapped GCC is propagated to user environments

### ISSUE THREAD

**Issue by abbradar** [NixOS/nixpkgs#21751]
## Issue description

Our CC wrapper doesn't wrap all the binaries that GCC provides -- namely prefixed with architecture `x86_64-unknown-gnu-gcc` and variations. Because unwrapped GCC is propagated to user environments from the wrapper it ends up in the path (for example in FHS environments). One example of problem that it causes is that OpenWrt picks it up and fails to compile.

`wrapCC` needs to detect and wrap those.

### Steps to reproduce

1. `buildEnv { paths = [ gcc ]; }` (you can also add it to `environment.systemPackages` or `buildFHSUserEnv`);
2. Try to compile something with `x86_64-unknown-gnu-gcc` (prefix may vary).

## Technical details

* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`) 59dbcefaa74b940f9b7490c034dc4ff885fa2627

**Comment by mmahut** [comment]
Are there any updates to this issue, please?

**Comment by collares** [comment]
Still relevant, I think. This might be interacting badly with newer versions of autoconf-archive macros AX_PROG_CC_FOR_BUILD and AX_PROG_CXX_FOR_BUILD, which are picking up unwrapped compilers. See https://github.com/NixOS/nixpkgs/pull/143521.

**Comment by L-as** [comment]
I think I've found a solution to this: https://github.com/NixOS/nixpkgs/pull/151357/commits/12cd91d691cb4acac4013666cb64a2e4079508ec#diff-1c3f080fce43dce74a94146b22cf2b6cb88f4a72c3491e82fef185393c152ff7R13

We just need `CC_FOR_BUILD=$CC`

**Comment by collares** [comment]
Thanks for taking a look at this!

Does this work for cross? If I understand correctly, `CC_FOR_BUILD` is supposed to be a compiler that produces binaries for the same architecture it runs on (that is, host == target for the compiler), while $CC is not if you're cross-compiling.

**Comment by L-as** [comment]
That's a good question. It *builds*, but I actually haven't tested the cross-compiled binaries. I *think* it works though.

**Comment by Ericson2314** [comment]
`CC_FOR_BUILD=$CC` is a non-starter for cross, I am afraid.

**Comment by Ericson2314** [comment]
@collares I am a bit confused. IMO the unwrapped binaries being used is the problem. Those should be pretty "bannished" from showing up in Nix builds.

**Comment by Ericson2314** [comment]
All this makes me wish more I had time to fix https://github.com/NixOS/nixpkgs/pull/132343. We can then simplify cc-wrapper, which will generally help restore sanity. We might also be able to use the resulting freed up "complexity budget" to make a `ccWithPacages` like `ghcWithPackages` which would be good for various ad-hoc compilation outside derivations.

**Comment by collares** [comment]
Here's `PATH` for the build in question (this is for the `singular` derivation). Note that the wrapped `gcc` appears just before the unwrapped `gcc`. The original comment in this issue says "[b]ecause unwrapped GCC is propagated to user environments from the wrapper", and I took it at face value. I will check if this is `singular`-specific, though.

```
PATH: /nix/store/mk6jf2n2bysmgvyl3g3krx1ml8r2b74z-bison-3.8.2/bin/
PATH: /nix/store/32n29c24bipfj2igyk8nbkvl8fpkfnm3-gnum4-1.4.19/bin/
PATH: /nix/store/8sfnian5c4kxrn7375cpcd5vq42gzfhy-perl-5.34.0/bin/
PATH: /nix/store/wsh0gbw514d4d996ynfgnjm82ki0fq1m-pkg-config-wrapper-0.29.2/bin/
PATH: /nix/store/l6f1d3jwcpnd6y7arxipxlcn9w8kxaqz-autoconf-2.71/bin/
PATH: /nix/store/k3hsxr3j3k8zmb04mdmm3hjbflawb73a-automake-1.16.3/bin/
PATH: /nix/store/zvvajyh85j0rbwkzfq5xddap762r4qy6-gettext-0.21/bin/
PATH: /nix/store/fbkyvpjnxhh7kkgv3jy9r88lj3xhf7y2-libtool-2.4.6/bin/
PATH: /nix/store/rhcg1qsv8p05cqd7am2az69hmi2i79sh-sharutils-4.15.2/bin/
PATH: /nix/store/hg41y6yp6rzl5z0fikn11hp9xva6dgll-doxygen-1.8.20/bin/
PATH: /nix/store/5801xhsdi62mir8012ma8si0ms3f9bl5-graphviz-2.49.3/bin/
PATH: /nix/store/lgqzk1naj9gq72zb164ggm5m158bmywl-texinfo-4.13a/bin/
PATH: /nix/store/m2qvdhzhfj7wdglyicaq1s8zr3biq3z7-texlive-combined-small-2021.20210408/bin/
PATH: /nix/store/18biibvrl9dg65ixz322x1wlccmgvkrl-patchelf-0.13/bin/
PATH: /nix/store/lmjarab5jydvmfng8zcqp40bcq0141z3-gcc-wrapper-9.3.0/bin/
PATH: /nix/store/919j2hfysp5iwvd2ssqkaj5n519pkaik-gcc-9.3.0/bin/
PATH: /nix/store/1d3vmlyd1c26gbdfjf14ykmm9qdmm094-glibc-2.33-56-bin/bin/
PATH: /nix/store/00b76mg3gjj4k394psj85gxmb047w69h-coreutils-9.0/bin/
PATH: /nix/store/r7izxhclvhvsjagssq0mnwwfrymlgzmg-binutils-wrapper-2.35.2/bin/
PATH: /nix/store/5rs0jfwgffx4cizpsn7ffhhisl22bxh0-binutils-2.35.2/bin/
PATH: /nix/store/x4kj84z42zm9r55h37qrny7b561h7wd1-ncurses-6.2-dev/bin/
PATH: /nix/store/8dcsbcip7n6s0lka6qjab1r2xvzwvh82-ncurses-6.2/bin/
PATH: /nix/store/hk73bm8ghyd0a4dr19fqc49k2i0l14qn-readline-6.3p08/bin/
PATH: /nix/store/vfn4gl8ss50cxp1g48345v7sibx4pjms-lrcalc-1.2/bin/
PATH: /nix/store/a05p4npf1i14qvsg61w13qhy5iwz7xhx-gfan-0.6.2/bin/
PATH: /nix/store/m3j7jccj9jja8wm0b5wlipwk8si9dzh7-cddlib-0.94m/bin/
PATH: /nix/store/00b76mg3gjj4k394psj85gxmb047w69h-coreutils-9.0/bin/
PATH: /nix/store/7pwbf3i7sbr839sjrpp3slj2xs3cqzag-findutils-4.8.0/bin/
PATH: /nix/store/cpinvxbv2w0w8sl3xnfikkjnvr71s993-diffutils-3.8/bin/
PATH: /nix/store/mfkb1fqhr43zhn251ynf7zmx52h0mg6d-gnused-4.8/bin/
PATH: /nix/store/inysrik3db3a125vb1ivwg5f9lprb9pi-gnugrep-3.7/bin/
PATH: /nix/store/0ixdpwkdn2rsx7algazgkgvm5gkj7nd8-gawk-5.1.1/bin/
PATH: /nix/store/pz324z68c586mm9kp22pcdxxar9igxqx-gnutar-1.34/bin/
PATH: /nix/store/dp647s0x345iczbwrlcqjmyabcpxq7hv-gzip-1.11/bin/
PATH: /nix/store/y7nh0cjqdz7qydanzinpavqal4jb8in4-bzip2-1.0.6.0.2-bin/bin/
PATH: /nix/store/7a65i3akpsiwjq9xb8rr26m1rkx4g4wj-gnumake-4.3/bin/
PATH: /nix/store/9b7h9qbf4c7c236f53nmrz3x0g9cfj0p-bash-5.1-p12/bin/
PATH: /nix/store/9jnf0kjfgnyinagi44jh52vn7x2ipdhn-patch-2.7.6/bin/
PATH: /nix/store/z8hqh6bayrx3d6id0frrcwjva7261gpm-xz-5.2.5-bin/bin/
PATH: /home/collares/.nix-profile/bin/
```

**Comment by Ericson2314** [comment]
@collares is your `config.log` snippet from a nix build or trying to do thing with compilers from the user environment? I am confused.

**Comment by collares** [comment]
The error happens with a nix build, but to access log file I did `nix develop nixpkgs#singular` and ran `genericBuild`. I should have created a new issue to avoid confusion, sorry about that! I don't think the user environment matters in my case, but I could be wrong, just the part about the unwrapped compilers being exposed by the wrapper. Let me know if there's any info I can provide!

**Comment by collares** [comment]
I can see the unwrapped version on the path of any derivation. Let me know if I'm doing something wrong.

```
$ nix-shell --pure -E 'with (import ./. {}); pkgs.hello'
this path will be fetched (0.69 MiB download, 0.69 MiB unpacked):
  /nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz
copying path '/nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz' from 'https://cache.nixos.org'...

[nix-shell:~/nixpkgs]$ echo $PATH
/nix/store/yx24h6ywbkg7rm5x7lrm8mx8ypcf4ywi-bash-interactive-5.1-p8/bin:/nix/store/pjba17qx8vh7dln8m5vnkqdbbjahvg9b-patchelf-0.13/bin:/nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/bin:/nix/store/iyssx2arl8bc40pjimyfxyknj06swjdx-gcc-10.3.0/bin:/nix/store/q6prgn66s4achzrrnvcvyjgnm94c1bn3-glibc-2.33-56-bin/bin:/nix/store/l6f4z8mmcnnxba8w004xn28y0vr4gdkf-coreutils-9.0/bin:/nix/store/v7w0pspwq8r2b7k2sndxq3db843z7xm5-binutils-wrapper-2.35.2/bin:/nix/store/lbxfixyw1yk099pjyaiy3xj5dl7kxm1g-binutils-2.35.2/bin:/nix/store/l6f4z8mmcnnxba8w004xn28y0vr4gdkf-coreutils-9.0/bin:/nix/store/v3lvq9hqshyldc4i6f5jy0zs0k5psbws-findutils-4.8.0/bin:/nix/store/gxj3wkwc5m5vb8b3rv6h029ky9nan4bf-diffutils-3.8/bin:/nix/store/vklvyr82ajbz7jm7g8dbkh62k20b0dpr-gnused-4.8/bin:/nix/store/nkwls56wcfwi1r0jnkqkvwx2zk7w3qrz-gnugrep-3.7/bin:/nix/store/21nzalbq7ay1vzanri1dl4845s3cl72i-gawk-5.1.1/bin:/nix/store/wny483nz9q52xq8azwpqna3pq98zcsgy-gnutar-1.34/bin:/nix/store/mp7lwrwr0nl69rznln13j3k4zf69s80i-gzip-1.11/bin:/nix/store/m9jxbbbxfzi6052yw3kgz95wvz76mbb1-bzip2-1.0.6.0.2-bin/bin:/nix/store/315glp1si526x60nhhkbqmip4m7b25a7-gnumake-4.3/bin:/nix/store/a54wrar1jym1d8yvlijq0l2gghmy8szz-bash-5.1-p12/bin:/nix/store/h2xzp4a764hn5n9br63ilh9qglz3baxn-patch-2.7.6/bin:/nix/store/3q4xj02ijqgfxffndvg28nlv741sx6qx-xz-5.2.5-bin/bin
```

**Comment by Ericson2314** [comment]
@collares so the weird thing is how does the autoconf macro somehow get the prefixed gcc which should be "shadowed" by the wrapper coming earlier in the PATH.

**Comment by collares** [comment]
There's no wrapper for the prefixed gcc here, just for the unprefixed one:

```
$ ls /nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/bin
as  c++  cc  cpp  g++  gcc  ld  ld.bfd  ld.gold  strip
```

**Comment by L-as** [comment]
I am not familiar with how autoconf works, but to me it looks like it tries looking for the compilers in `$PATH` in alphabetic order, meaning `aarch64...` comes before `cc` or `gcc`. The same does not hold for `x86...`.
Of course this may be very wrong.

**Comment by L-as** [comment]
Ah, it's probably what collares says.

**Comment by Ericson2314** [comment]
> There's no wrapper for the prefixed gcc here, just for the unprefixed one

Why is this? We should always be using wrapped compilers in nix builds.

**Comment by collares** [comment]
I don't think I have any good explanation here, but I would be happy to run any tests you think would be useful. Three things seem to be happening:

* My stdenv does not contain wrappers whose names contain prefixes. This is also the case on the aarch64 build that failed.
* The directory containing the unwrapped binaries ends up in PATH (after the wrappers, but still)
* The newer version of the `AX_PROG_CC_FOR_BUILD` macro guesses a prefixed binary name (`aarch64-unknown-linux-gnu-gcc`) and the only thing on the PATH matching that name is the unwrapped version.

The end result is that the macro ends up finding an unwrapped compiler. The macro was only updated last year, and the first version of `autoconf-archive` that contains the updated version was released this year, so not a lot of packages have updated yet, and the number of packages that use `CC_FOR_BUILD` is probably not too large.

**Comment by collares** [comment]
After reading https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/cc-wrapper/default.nix#L36, I think I see a possible source of confusion. To clarify, my comment above (https://github.com/NixOS/nixpkgs/issues/21751#issuecomment-997471353) does not refer to a cross build: The build that failed on aarch64 was supposed to generate aarch64 binaries.

**Comment by Ericson2314** [comment]
Hmm so the issue is that the native unwrapped binaries are both prefixed and unprefixed? CC @sternenseemann who was looking down exactly when GCC is prefixed to help dodge this sort of thing.

**Comment by collares** [comment]
Yep, that's the issue! This wasn't a problem before, but with the new autoconf macros it is.

**Comment by sternenseemann** [comment]
We force the prefix to be the same as what nix thinks `targetPrefix` should be in general. So I guess we need to figure out why `CC_FOR_BUILD` would get a prefix for a native build?

**Comment by Ericson2314** [comment]
@sternenseemann I am not sure CC_FOR_BUILD is set at all in this case, I think autoconf is just guessing names, tries with the prefix first, and then slips passed the wrapped one to find the unwrapped one. But I am not sure.

**Comment by collares** [comment]
Yes, exactly, `CC_FOR_BUILD` is unset on the corresponding `nix develop` shell, since the derivation didn't have the native compiler in `depsBuildBuild`. Arguably this is a problem with the derivation, since `depsBuildBuild` should be set if the build will require something that is definitely a native compiler. But the fact that post-https://github.com/autoconf-archive/autoconf-archive/pull/207 autoconf finds the prefixed (and unwrapped) version automatically on a native build and then fails to generate executables with it leads to very confusing error messages, especially since things work fine on x86_64 (autoconf guesses the wrong prefix there).

**Comment by collares** [comment]
If it helps, I have confirmed that https://github.com/autoconf-archive/autoconf-archive/pull/207/commits/9c26407efb0322dda2c08fada6a67dde1c690e74 is the particular commit that causes the `AX_PROG_CC_FOR_BUILD` macro to guess the prefixed name for [this configure.ac](https://github.com/Singular/Singular/blob/0c26a150cc7724e89eac99e4b5d787d781c422f0/omalloc/configure.ac).

**Comment by Ericson2314** [comment]
Yeah I put no blame on autoconf. We simple should make sure that no compiler, unprefixed otherwise, can 
"leak" past the wrapper earlier the PATH.

Put simply, if the unwrapped provides prefixed, the wrapped should provide prefixed, and if the unwrapped provides unprefixed, the wrapped should also provide unprefixed.

**Comment by sternenseemann** [comment]
I think we just need to prevent `gcc-unwrapped` from creating prefixed compilers if `--program-prefix=""`. `targetPrefix` should be accurate and the only way to find the compiler. It seems like as a quick fix we could delete all symlinks in gcc's `$out/bin`, although I'm a bit confused by `${stdenv.targetPlatform.config}-gcc-{ar, nm, ...}` which seem to be proper binaries?

**Comment by Ericson2314** [comment]
That makes sense, thanks for taking a look. I guess we can rename those but also see why they were make like that in the first place.

**Comment by sternenseemann** [comment]
It creates both `gcc-ar` and `x86_64-unknown-linux-gcc-ar` for some reason, whereas `x86_64-unknown-linux-gcc` is just a symlink to `gcc`.

**Comment by collares** [comment]
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/gcc/builder.sh#L279
